go/inst/instance_dao.go

```diff
@@ -1738,42 +1750,64 @@ func GetClusterOSCReplicas(clusterName string) ([](*Instance), error) {
 		if err != nil {
 			return result, err
 		}
-		sort.Sort(sort.Reverse(InstancesByCountReplicas(intermediateMasters)))
+		// intermediateMasters has all IMs
 		intermediateMasters = filterOSCInstances(intermediateMasters)
-		intermediateMasters = intermediateMasters[0:math.MinInt(2, len(intermediateMasters))]
-		result = append(result, intermediateMasters...)
+		/*
+			fmt.Println("All IMs: ")
+			for _, im := range intermediateMasters {
+				fmt.Println(im.Key)
+			}
+		*/
+		// sort by DC and replicas count
+		sort.Sort(sort.Reverse(InstancesByDc(intermediateMasters)))
+		/*
+			fmt.Println("All IMs sorted by Dc: ")
+			for _, im := range intermediateMasters {
+				fmt.Println(im.Key, im.DataCenter, len(im.Replicas))
+			}
+		*/
+		// pick 2 busiest from every DC
+		currentDCInstances := 0
+		var currentDC *string = nil
+
+		for _, im := range intermediateMasters {
+			if currentDC == nil || *currentDC != im.DataCenter {
+				currentDCInstances = 0
+				currentDC = &im.DataCenter
+				//				fmt.Println("Examining DataCenter: ", im.DataCenter)
+			}
+			if currentDCInstances > 1 {
+				continue
+			}
+			currentDCInstances++
+			result = append(result, im)
+			//			fmt.Println("Appending IM: ", im.Key)
+		}
 	}
 	{
-		// Get 2 replicas of found IMs, if possible
-		if len(intermediateMasters) == 1 {
-			// Pick 2 replicas for this IM
-			replicas, err := ReadReplicaInstances(&(intermediateMasters[0].Key))
+		// Get at most 2 replicas of found IMs
+		for _, im := range result {
+			replicas, err := ReadReplicaInstances(&im.Key)
 			if err != nil {
 				return result, err
 			}
+			// todo: we get most busy replicas, but wouldn't it make sense
+			// to use most lagging ones?
 			sort.Sort(sort.Reverse(InstancesByCountReplicas(replicas)))
 			replicas = filterOSCInstances(replicas)
 			replicas = replicas[0:math.MinInt(2, len(replicas))]
 			result = append(result, replicas...)
-
-		}
-		if len(intermediateMasters) == 2 {
-			// Pick one replica from each IM (should be possible)
-			for _, im := range intermediateMasters {
-				replicas, err := ReadReplicaInstances(&im.Key)
-				if err != nil {
-					return result, err
+			/*
+				fmt.Println("Appending replicas: ")
+				for _, rep := range replicas {
+					fmt.Println(rep.Key)
 				}
-				sort.Sort(sort.Reverse(InstancesByCountReplicas(replicas)))
-				replicas = filterOSCInstances(replicas)
-				if len(replicas) > 0 {
-					result = append(result, replicas[0])
-				}
-			}
+			*/
 		}
 	}
 	{
 		// Get 2 3rd tier replicas, if possible
+		// todo: again, we get most busy. Why not most laging?
```

Comment:
typo: lagging
- reasoning: you'd need to ask Shlomi
- I think that the IMs with most replicas are most likely to busier and thus affected by replication delay than general slaves, so the approximation while not perfect is not bad. Is it fully correct? Maybe not.